### ISSUE TITLE
 Disable Default CPU Limits for Istio Proxy-Version 1.11.3

### ISSUE THREAD

**Issue by Mohid-A** [istio/istio#40078]
### Bug Description

Removing the CPU limits for the istio sidecar is not working and adds the default CPU limit of **2000m** to the sidecar. The default is causing the cluster to go into the overcommitted state, when the CPU limits are enforced based on the usage trend it's resulting in CPU throttling causing performance issues. Tried setting the limits value to “**null”** but it did not change the behavior. Setup is using the operator `apiVersion: install.istio.io/v1alpha1
kind: IstioOperator` and the CPU/memory values are updated under the global config as shown in the below snippet

```

`      proxy:
         resources:
          limits:
            memory: 150Mi
          requests:
            cpu: 20m
            memory: 150Mi
        statusPort: 15020
        tracer: zipkin
       proxy_init:
        image: proxyv2
        resources:
          limits:
            memory: 150Mi
          requests:
            cpu: 20m
            memory: 150Mi`

```

### Version

```prose
client version: 1.11.3
control-plane version: 1.11.3
data plane version: 1.11.3 (64 proxies)
```

### Additional Information

Output for >k get po sample-schema-producer -o yaml ; shows the default limit of 2 CPU enforced on the container

```
`    image: docker.io/istio/proxyv2:1.11.3
     imagePullPolicy: IfNotPresent
     name: istio-proxy
     ports:
     - containerPort: 15090
        name: http-envoy-prom
        protocol: TCP
     readinessProbe:
       failureThreshold: 30
       httpGet:
        path: /healthz/ready
        port: 15021
        scheme: HTTP
     initialDelaySeconds: 20
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 3
    resources:
      limits:
        cpu: "2"
        memory: 150Mi
      requests:
        cpu: 20m
        memory: 150Mi`
```

**Comment by chargui-yassine** [comment]
Same issue with version 1.13.9. Do we have any fix for this?

**Comment by mario-shippo** [comment]
Could someone help identify if I'm missing anything here? 
The issue was previously closed, citing this [merge](https://github.com/deckhouse/deckhouse/pull/2579) as a solution. Despite implementing the patch according to the unit test changes, **without success**, any modification to the Operator results in limits.cpu resetting to 2. 
It is a long running issue....and significant problem, potentially causing CPU throttling in pods and leading to latency issues.

**Comment by binboum** [comment]
I'm writing this brief message for those who might be facing this issue and haven't found a solution yet. Historically, in the past, I used the "null" value for the CPU limit, which effectively overloaded it. However, this method doesn't seem to work, at least not with the Helm method.

Based on my initial research, for those using Helm to deploy (istiod), the default values are set to 2.
I think it doesn't override the existing value. There's also this issue with Helm that could be part of the answer: 

In any case, if you take a closer look at the configmap for the injector : 

https://github.com/helm/helm/issues/12421
https://github.com/istio/istio/blob/master/manifests/charts/istio-control/istio-discovery/values.yaml#L351
```
data:
  config: |-
    # defaultTemplates defines the default template to use for pods that do not explicitly specify a template
    defaultTemplates: [sidecar]
    policy: enabled
    alwaysInjectSelector:
      []
    neverInjectSelector:
      []
    injectedAnnotations:
      "cluster-autoscaler.kubernetes.io/safe-to-evict": "true"
    template: "{{ Template_Version_And_Istio_Version_Mismatched_Check_Installation }}"
    templates:
      sidecar: |
        {{- define "resources"  }}
          {{- if or (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyCPU`) (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyMemory`) (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyCPULimit`) (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyMemoryLimit`) }}
            {{- if or (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyCPU`) (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyMemory`) }}
              requests:
                {{ if (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyCPU`) -}}
                cpu: "{{ index .ObjectMeta.Annotations `sidecar.istio.io/proxyCPU` }}"
                {{ end }}
                {{ if (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyMemory`) -}}
                memory: "{{ index .ObjectMeta.Annotations `sidecar.istio.io/proxyMemory` }}"
                {{ end }}
            {{- end }}
            {{- if or (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyCPULimit`) (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyMemoryLimit`) }}
              limits:
                {{ if (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyCPULimit`) -}}
                cpu: "{{ index .ObjectMeta.Annotations `sidecar.istio.io/proxyCPULimit` }}"
                {{ end }}
                {{ if (isset .ObjectMeta.Annotations `sidecar.istio.io/proxyMemoryLimit`) -}}
                memory: "{{ index .ObjectMeta.Annotations `sidecar.istio.io/proxyMemoryLimit` }}"
                {{ end }}
            {{- end }}
          {{- else }}
            {{- if .Values.global.proxy.resources }}
              {{ toYaml .Values.global.proxy.resources | indent 6 }}
            {{- end }}
          {{- end }}
        {{- end }}
```

You'll notice that using annotations allows you to bypass the default parameters, making it a functional workaround. 
I'll try to take the time to verify my leads. In the meantime, you can use this approach.

**Comment by mario-shippo** [comment]
For those still facing issues with that, the best alternative I could find until now was to set cpu to 0 (zero).
Anything else wont work or will required to define annotations to all pods which can be challenge in a mixed env.
This is from test env, so be carefull with any configurations below:

`apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  annotations:
  name: istio-operator
  namespace: istio-system
spec:
  components:
    base:
      enabled: true
    cni:
      enabled: false
    egressGateways:
      - enabled: true
        k8s:
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                  - matchExpressions:
                      - key: EGRESS_ALLOWED
                        operator: In
                        values:
                          - 'true'
          hpaSpec:
            maxReplicas: 1
            minReplicas: 1
          resources:
            limits:
              cpu: '0'
              memory: 250Mi
            requests:
              cpu: '0'
              memory: 128Mi
          tolerations:
            - effect: NoExecute
              key: EGRESS_ALLOWED
              operator: Exists
        name: istio-egressgateway
    ingressGateways:
      - enabled: true
        k8s:
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                  - matchExpressions:
                      - key: INGRESS_ALLOWED
                        operator: In
                        values:
                          - 'true'
          env:
            - name: ISTIO_META_HTTP10
              value: '1'
            - name: WASM_INSECURE_REGISTRIES
              value: '*'
          hpaSpec:
            maxReplicas: 1
            minReplicas: 1
          resources:
            limits:
              cpu: '0'
              memory: 250Mi
            requests:
              cpu: '0'
              memory: 128Mi
          service:
            externalTrafficPolicy: Local
            loadBalancerIP: 123.456.789.012
            type: LoadBalancer
          tolerations:
            - effect: NoExecute
              operator: Exists
        name: istio-ingressgateway
    istiodRemote:
      enabled: false
    pilot:
      enabled: true
      k8s:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: ISTIO_ALLOWED
                      operator: In
                      values:
                        - 'true'
        env:
          - name: WASM_INSECURE_REGISTRIES
            value: '*'
        hpaSpec:
          maxReplicas: 1
          minReplicas: 1
        resources:
          limits:
            cpu: '0'
            memory: 1Gi
          requests:
            cpu: '0'
            memory: 128Mi
        tolerations:
          - effect: NoExecute
            operator: Exists
  meshConfig: {}
  profile: default
  values:
    global:
      proxy:
        resources:
          limits:
            cpu: '0'
            memory: 150Mi
          requests:
            cpu: '0'
            memory: 50Mi
    sidecarInjectorWebhook:
      neverInjectSelector:
        - matchExpressions:
            - key: type
              operator: In
              values:
                - job
                - cron
        - matchExpressions:
            - key: layer
              operator: In
              values:
                - plugins
        - matchExpressions:
            - key: job-name
              operator: Exists
`

**Comment by martijnhybrit** [comment]
I'm not sure why this is closed, I'm still experiencing this issue with v1.24.1.
I'm working on a cluster that has Kyverno as a policy engine which has a policy that forbids the setting of CPU limits. This includes setting it to 0. The istio proxy is triggering this as it refuses to take `null` as a setting:

```
global:
  platform: openshift
  proxy:
    resources:
      requests:
        cpu: 10m
        memory: 128Mi
      limits:
        cpu: null # this is just ignored
        memory: 512Mi

```

**Comment by afrancis101** [comment]
Was able to resolve this by setting `defaults.global.proxy.resources.limits.cpu` to null instead of `global.proxy.resources.limits.cpu`
```
# Setting `defaults.global.proxy.resources.limits.cpu: null` - snippet from istio-sidecar-injector configmap
"resources": {
  "limits": {
    "memory": "1024Mi"
  },
  "requests": {
    "cpu": "100m",
    "memory": "128Mi"
  }
},

# Setting `global.proxy.resources.limits.cpu: null` - snippet from istio-sidecar-injector configmap
"resources": {
  "limits": {
    "cpu": null,
    "memory": "1024Mi"
  },
  "requests": {
    "cpu": "100m",
    "memory": "128Mi"
  }
},
```

**Comment by Westacular** [comment]
To anyone who was confused (as I was) about the inconsistent reports of whether using `cpu: null` would work to remove the limit:

- Using it with `kubectl apply` works
- However, if you use Helm, there was a longstanding bug in Helm related to coalescing sub-charts that meant that `null` values were being ignored and were _not_ removing a default value when they were expected to: https://github.com/helm/helm/issues/5184 https://github.com/helm/helm/issues/12637
- the Helm bug was fixed in v3.17.1 about a year ago

**Comment by frittentheke** [comment]
Maybe also related https://github.com/kubernetes/kubernetes/issues/135423
